File: highwaycombinatorics/submissions/accepted/paul.cpp
#include <bits/stdc++.h>
using namespace std;
using i64 = long long;

const int N = 85, MOD = 1e9 + 7;
int fib[N];

i64 exp_mod(i64 a, i64 n) {
	i64 res = 1;
	for (; n > 0; n /= 2) {
		if (n%2) res = res*a % MOD;
		a = a*a % MOD;
	}
	return res;
}


i64 inv_mod(i64 a) {
	return exp_mod(a, MOD-2);
}

unordered_set<i64> values;
void gen_partitions(int rem, int amax, i64 x) {
	values.insert(x);
	for (int a = 2; a <= amax && a < rem; a++) {
		gen_partitions(rem-a-1, a, x*fib[a] % MOD);
	}
}

vector<i64> fail = {-1};
vector<i64> reconstruct(i64 y, int rem = N, int amax = N, i64 x = 1) {
	if (x == y) return {};
	for (int a = 2; a <= amax && a < rem; a++) {
		auto v = reconstruct(y, rem-a-1, a, x*fib[a] % MOD);
		if (v != fail) { v.push_back(a); return v; }
	}
	return fail;
}

int main() {
	fib[0] = fib[1] = 1;
	for (int n = 2; n < N; n++) {
		fib[n] = (fib[n-1] + fib[n-2]) % MOD;
	}
	gen_partitions(N, N, 1);
	cerr << values.size() << '\n';

	i64 x; cin >> x;

	if (x == 0) {
		cout << ".##" << '\n' << "##." << '\n';
	} else {
		for (i64 y: values) {
			i64 z = x * inv_mod(y) % MOD;
			if (values.count(z)) {
				string res = "#";
				for (int k: reconstruct(y)) {
					res += string(k, '.') + "#";
				}
				for (int k: reconstruct(z)) {
					res += string(k, '.') + "#";
				}
				cout << res << '\n' << res << '\n';
				break;
			}
		}
	}
}

File: highwaycombinatorics/submissions/accepted/paul.py
#!/usr/bin/env python3

N, p = 83, 10**9 + 7

fib = [1, 1]
while len(fib) < N:
  fib.append((fib[-1]+fib[-2]) % p)

values = set()
def gen_partitions(rem, amax, x):
  values.add(x)
  for a in range(2, min(amax+1, rem)):
    gen_partitions(rem-a-1, a, x*fib[a] % p)

gen_partitions(N, N, 1)

def reconstruct(rem, amax, x, y):
  if x == y:
    return []
  for a in range(2, min(amax+1, rem)):
    v = reconstruct(rem-a-1, a, x*fib[a] % p, y)
    if v is not None:
      return v + [a]

x = int(input())
if x == 0:
  print('.##\n##.')
else:
  for y in values:
    z = x * pow(y, p-2, p) % p
    if z in values:
      v = reconstruct(N, N, 1, y) + reconstruct(N, N, 1, z)
      res = '#'
      for k in v:
        res += '.' * k + '#'
      print(res)
      print(res)
      break
